File: fsharp/log-levels/LogLevels.fs
module LogLevels

let (|Prefix|_|) (p: string) (s: string) =
    if s.StartsWith(p) then
        Some(s.Substring(p.Length).Trim())
    else
        None

let message (logLine: string) : string =
    match logLine with
    | Prefix "[ERROR]: " logMsg -> logMsg
    | Prefix "[WARNING]: " logMsg -> logMsg
    | Prefix "[INFO]: " logMsg -> logMsg
    | _ -> ""


let logLevel (logLine: string) : string =
    match logLine with
    | Prefix "[ERROR]: " _ -> "error"
    | Prefix "[WARNING]: " _ -> "warning"
    | Prefix "[INFO]: " _ -> "info"
    | _ -> ""

let reformat (logLine: string) : string =
    match logLine with
    | Prefix "[ERROR]: " logMsg -> logMsg + " (error)"
    | Prefix "[WARNING]: " logMsg -> logMsg + " (warning)"
    | Prefix "[INFO]: " logMsg -> logMsg + " (info)"
    | _ -> ""

File: csharp/log-levels/LogLevels.cs
using System;
using System.Runtime.CompilerServices;

static class LogLine
{
    public static string Message(string logLine)
    {
        var message = logLine.Split("]:");
        return message[1].Trim();
    }

    public static string LogLevel(string logLine)
    {
        var logLevel = logLine.Split("]:");
        return logLevel[0].Trim()[1..].ToLower();
    }

    public static string Reformat(string logLine)
    {
        var message = Message(logLine);
        var logLevel = LogLevel(logLine);

        return $"{message} ({logLevel})";
    }
}
